torch/nn/utils/scale_grad.py

```diff
@@ -0,0 +1,80 @@
+# mypy: allow-untyped-decorators
+from collections import defaultdict
+from typing import Optional
+
+import torch
+from torch import Tensor
+from torch.nn.utils.clip_grad import _no_grad, _tensor_or_tensors
+from torch.utils._foreach_utils import _device_has_foreach_support, _has_foreach_support
+
+
+__all__ = [
+    "scale_grad_",
+]
+
+
+@_no_grad
+def scale_grad_(
+    parameters: _tensor_or_tensors,
+    scaler: torch.Tensor,
+    foreach: Optional[bool] = None,
+) -> None:
+    r"""Scale gradients of iterable parameters.
+
+    This function is equivalent to :func:`torch.mul_` applied to each parameter.
+    Gradients are modified in-place, multiplying by specified scaler.
+
+    Args:
+        parameters (Iterable[Tensor] or Tensor): an iterable of Tensors or a
+            single Tensor that will have gradients scaled
+        scaler (Tensor): multiplier to scale gradients
+        foreach (bool): use the faster foreach-based implementation.
+            If ``None``, use the foreach implementation for CUDA and CPU native tensors and silently
+            fall back to the slow implementation for other device types.
+            Default: ``None``
+    Returns:
+        None
+    """
+    if isinstance(parameters, torch.Tensor):
+        parameters = [parameters]
+    else:
+        parameters = list(parameters)
+    _scale_grad_(parameters, scaler, foreach)
+
+
+def _group_tensors_by_device_and_dtype(
```

Comment:
why redefine this vs use the C++ version?

Reply:
cc @mlazos who handled this back when working on optimizers--if i remember correctly, it was faster to have the compiler do the grouping?

Reply:
sorry I missed this, yes I had dynamo do this during tracing.